cmake/modules/RootMacros.cmake

```diff
@@ -873,29 +1022,29 @@ function(ROOT_ADD_INCLUDE_DIRECTORIES library)
       # get a redefinition error.
       # We should remove these lines when the fallback include is removed. Then
       # we will need a module.modulemap file per `inc` directory.
-      target_include_directories(${library} BEFORE PRIVATE ${CMAKE_BINARY_DIR}/include)
+      # target_include_directories(${library} PRIVATE ${CMAKE_BINARY_DIR}/include)
     else()
       # needed for generated headers like RConfigure.h and ROOT/RConfig.hxx
-      target_include_directories(${library} PRIVATE ${CMAKE_BINARY_DIR}/include)
+      # target_include_directories(${library} PRIVATE ${CMAKE_BINARY_DIR}/include)
     endif()
 
   endif()
 
 endfunction(ROOT_ADD_INCLUDE_DIRECTORIES)
 
 #---------------------------------------------------------------------------------------------------
-#---ROOT_OBJECT_LIBRARY( <name> source1 source2 ... BUILTINS dep1 dep2 ...)
+#---ROOT_OBJECT_LIBRARY( <name> source1 source2 ... BUILTINS dep1 dep2 ... INCLUDES src1 src2)
 #---------------------------------------------------------------------------------------------------
 function(ROOT_OBJECT_LIBRARY library)
-  CMAKE_PARSE_ARGUMENTS(ARG "" "" "BUILTINS"  ${ARGN})
+  CMAKE_PARSE_ARGUMENTS(ARG "" "" "BUILTINS;INCLUDES"  ${ARGN})
   ROOT_GET_SOURCES(lib_srcs src ${ARG_UNPARSED_ARGUMENTS})
   add_library( ${library} OBJECT ${lib_srcs})
   if(lib_srcs MATCHES "(^|/)(G__[^.]*)[.]cxx.*")
      add_dependencies(${library} ${CMAKE_MATCH_2})
   endif()
   add_dependencies(${library} move_headers)
 
-  ROOT_ADD_INCLUDE_DIRECTORIES(${library})
+  ROOT_ADD_INCLUDE_DIRECTORIES(${library} INCLUDES ${ARG_INCLUDES})
```

Comment:
I'd prefer
```
ROOT_OBJECT_LIBRARY(foo foo.cxx)
target_include_directories(foo somedir)
```
instead of passing having to pass everything around. See also https://cliutils.gitlab.io/modern-cmake/chapters/basics.html "Targets are your friend"

Reply:
No, it does not work this way.
I have to use `ROOT_ADD_INCLUDE_DIRECTORIES` where all library dependency resolved and I collect all necessary includes. Function used in three places - I have to copy and paste same code three times

Reply:
Here how I implement it: https://github.com/root-project/root/pull/5181/commits/c1cb5863229ce52b1ae0b23f13972141ee448b3d